dashboard/app/seeding/lesson_seeding.rb

```diff
@@ -0,0 +1,69 @@
+require 'script_lessons_serializer'
```

Comment:
> I was unsure about where to best put the new module, but based on some quick googling I chose to make a new directory under app, rather than in lib, since this depends heavily on our specific business logic. Most of our misc stuff is already in lib, though, so I'd be ok with moving it there. Thoughts?

Good question :-) It's a bit hard to say without being hooked up to the code that will call it. Where will this module be included? 

My first thought is that the process of serializing and seeding seems closely tied to the model and should live in the Lesson model. One argument to support this is in https://codeclimate.com/blog/7-ways-to-decompose-fat-activerecord-models/ which says "Don’t Extract Mixins from Fat Models", and this implementation seems like it might meet that description -- it tries to slice of functionality that is specific to Lesson that is related to the "concern" of serialization/seeding. (NOTE: I/we have been bad about this, which has resulted in things like `UserPermissionGrantee` being extracted which violates this rule. so I'm open to not adopting this piece of advice)

Some thinking around this problem has been captured by @Hamms in https://github.com/code-dot-org/code-dot-org/blob/staging/dashboard/lib/services/README.md, which links to a list of articles in https://github.com/code-dot-org/code-dot-org/pull/30444 . Are you going off at the same list of articles? and/or are there others that you would add to that list?

What are your thoughts, @Hamms ?

Reply:
I actually think this is a fantastic candidate for a Service Object! It's super easy to think of this functionality as being specific to Lessons because we've so closely associated it with that specific model in the past, but really our seeding process needs to be aware of several different models in our system as it currently stands, and I fully expect that as we continue to develop out our broader level hierarchy we're going to want to add even more models into our system.

I'd love to see a Service Object or even a small family of Service Objects that can fully own the entire seeding process for all our content, and for those to all live in `dashboard/lib/services`

Reply:
Thanks @Hamms ! Just confirming a shared understanding that this includes not just seeding, but also serializing, which is done when a levelbuilder makes edits. the serializer method `serialize_lessons` is just a thin wrapper on the activemodel serializer. should that be separate?

Reply:
oh, good callout! I think I'd probably lean toward putting that wrapper somewhere else, yes. My thinking here is that activemodel serializers explicitly don't handle deserialization (as far as I'm aware), so our deserializer/seeder should likewise be decoupled from the serializer.

Would it work to put `serialize_lessons` as a class method on the serializer itself?

Reply:
> "Don’t Extract Mixins from Fat Models"

I'm not sure if that applies here, since this module isn't a Mixin?

From that same article, debatably, #2 might be an argument in favor of extraction. Bolded the points that might apply

> 2. Extract Service Objects
Some actions in a system warrant a Service Object to encapsulate their operation. I reach for Service Objects when an action meets one or more of these criteria:
>- **The action is complex (e.g. closing the books at the end of an accounting period)**
>- **The action reaches across multiple models (e.g. an e-commerce purchase using Order, CreditCard and Customer objects)**
>- The action interacts with an external service (e.g. posting to social networks)
>- **The action is not a core concern of the underlying model (e.g. sweeping up outdated data after a certain time period).**
>- There are multiple ways of performing the action (e.g. authenticating with an access token or password). This is the Gang of Four Strategy pattern.

As for the list of articles, I wasn't going off them specifically, though I can read through them now. I think wanting to break this out into a separate file was just a general programming feeling, and I did a tiny bit of ad-hoc googling to for `app` vs `lib`. But just going from general programming thoughts:

- In general, files that are too big get unwieldy - harder to find things in them by scrolling through, harder to remember everything that's in them, they even load slowly in your IDE
- This logic is makes sense to go together logically (it's all serialization and seeding), and only needs to be called in a small number of specific places. It's not something that we'll be constantly want to be invoking on arbitrary objects in arbitrary places.

Another analogy to something in our codebase could be `ScriptDSL` - it's doing the same kind of thing this is, right? But we probably wouldn't want to fold it into the Script model.

Reply:
All makes sense. I am on board with moving the seeding logic into a Service object!

I am not sure I understand the difference between a mixin and a module that gets included into a model. I thought that was a mixin? anyway, I feel better about it if the seeding logic is wrapped in a class instead, since that seems better encapsulated than an included module.

Reply:
> a mixin and a module that gets included into a model

I'm not that familiar, but I think you're right that these are the same. I'm not planning to do that with this new module. I'm planning to consume it like the `LessonSeedingTest` unit tests do, something like:

```
require 'lesson_seeding'

...

LessonSeeding.serialize_script_lessons(script)
```

Something that would work in any language, no Ruby weirdness.

Reply:
> oh, good callout! I think I'd probably lean toward putting that wrapper somewhere else, yes. My thinking here is that activemodel serializers explicitly don't handle deserialization (as far as I'm aware), so our deserializer/seeder should likewise be decoupled from the serializer.
>
>Would it work to put serialize_lessons as a class method on the serializer itself?

@Hamms This seems reasonable, but my thinking here was that this specific serializer is intended only to be used for this specific seeding logic, so those two should be coupled. One way to make this clearer could be to move those Serializer classes into this same module, and leave `serialize_lessons` where it is. How does that sound?

On `app` vs `lib` - there are a lot of articles online that suggest that things that are very domain-specific shouldn't go into `lib`. For example: https://devblast.com/b/rails-app-vs-lib

I ultimately don't care too much about this and am happy to follow our existing patterns, but I was curious. Thoughts?

Reply:
> I'm planning to consume it like the `LessonSeedingTest` unit tests do, something like:
> ```
> LessonSeeding.serialize_script_lessons(script)
> ```
> Something that would work in any language, no Ruby weirdness.

I think we're converging on the same solution here... but just wanted to suggest that if you're going to call the methods that way, then it would be better as a class than a module.

Reply:
> On `app` vs `lib` - there are a lot of articles online that suggest that things that are very domain-specific shouldn't go into `lib`. For example: https://devblast.com/b/rails-app-vs-lib

the following jumps out at me from that article:

> The /lib folder is for **libraries that are not packaged as gems but could be**.

This seems like a much more restrictive idea of what belongs in lib/ than some of the other articles we've been going off of. I can't really say that one or the other is better. The most important thing seems to be to move functionality out of models, views, controllers, etc. It seems like rails doesn't really provide any good convention for these kind of extractions so people have built their own strategies around the directories that do exist.

However, we seem to have some momentum around what's described in dashboard/lib/services which is a bit more inclusive as it includes business logic which by definition could not be a gem, since business logic is specific to the application. So if there is no strong preference otherwise, I'm inclined to go with what's documented there and put this in services.

One last consideration is that i noticed that app/ and lib/ are both in the "load path"... technically we don't need the seeding code to be loaded by the app, since we only need it during seed. so if we thought that was slowing down our app boot time, that could be something that we put outside of directories that are in the load path. I don't want to randomize us too much though so maybe this could be something we try to remember to "optimize later".